### ISSUE TITLE
High level Plan for fine-tuning once the data is ready

### ISSUE THREAD

**Issue by NisithDivantha** [psychlone77/fyp-ai4db#3]
To beat their performance without increasing model size (staying in the ~7B-8B parameter range), you must change how the model learns. E2ETUNE used standard Supervised Fine-Tuning (SFT), essentially forcing the model to memorize the mapping.

Here is the 3-step strategy to outperform E2ETUNE using 2025-era techniques: Better Base Model, Reasoning (CoT), and Preference Optimization (DPO).

1. Swap the Base Model (Immediate Gain)
E2ETUNE used Mistral-7B-v0.1 (2023). In late 2024/2025, significantly stronger models exist at the same size.

Recommendation: Qwen2.5-Coder-7B-Instruct.

Why? This model is "State-of-the-Art" for its size in code generation and structured output (JSON). Since database tuning is effectively a "text-to-JSON" coding task, Qwen2.5-Coder consistently outperforms general-purpose models like Mistral 7B or Llama 3.1 on these tasks.

Alternative: Llama-3.1-8B-Instruct. It has slightly more knowledge but is less specialized in structured output than Qwen Coder.

2. Strategy A: "Chain-of-Thought" Fine-Tuning
E2ETUNE maps Workload -> Config. This is hard because the model has to "guess" the logic implicitly. You will map Workload -> Reasoning Trace -> Config.

The Process:

Data Augmentation (Teacher Step): Take your labeled pairs (Workload, Config). Use a smart model (like GPT-4o or Claude 3.5 Sonnet) to generate the missing middle part.

Prompt: "Here is a workload [X] and the optimal DB config [Y]. Explain step-by-step using PostgreSQL internal logic WHY this config is optimal for this workload."

New Dataset Structure: Create a training file where the target output includes the reasoning before the JSON.

JSON

{
  "input": "Workload: Read-heavy, 50GB dataset...",
  "output": "Based on the read-heavy nature... [Reasoning]... \n ```json { \"shared_buffers\": \"16GB\" ... } ```"
}
Why this beats E2ETUNE: The model learns the causal link between workload features (e.g., "many sort operations") and knob values (e.g., "increase work_mem"), allowing it to generalize to unseen workloads much better.

3. Strategy B: Direct Preference Optimization (DPO)
This is the "secret weapon" that E2ETUNE did not use. SFT only teaches the model to imitate a good config. DPO teaches the model to rank a good config higher than a bad one.

You already have the data for this. When you ran HEBO to label your dataset, you didn't just find the best config; you found many bad ones too.

The Process:

Create Triplets: From your HEBO logs, extract pairs for the same workload:

Chosen: The config that gave the lowest latency.

Rejected: A config that gave higher latency (e.g., +20% slower).

Train with DPO: Use the TRL (Transformer Reinforcement Learning) library to fine-tune your SFT model using these pairs.

Objective: The loss function penalizes the model if it assigns high probability to the "Rejected" config.

Why this beats E2ETUNE: It explicitly aligns the model with your true metric (latency/performance) rather than just text similarity.
